pkg/cloud/api/obj.go

```diff
@@ -0,0 +1,300 @@
+/*
+Copyright 2023 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// ConversionContext gives which version => version the error occurred on.
+type ConversionContext int
+
+const (
+	GAToAlphaConversion ConversionContext = iota
+	GAToBetaConversion
+	AlphaToGAConversion
+	AlphaToBetaConversion
+	BetaToGAConversion
+	BetaToAlphaConversion
+	conversionContextCount
+)
+
+// ConversionError is returned from To*() methods. Inspect this error to get
+// more details on what did not convert.
+type ConversionError struct {
+	// MissingFields is a list of field values that were set but did not
+	// translate to the version requested.
+	MissingFields []MissingField
+}
+
+func (e *ConversionError) hasErr() bool {
+	return len(e.MissingFields) > 0
+}
+
+func (e *ConversionError) Error() string {
+	return fmt.Sprintf("ConversionError: missing fields %v", e.MissingFields)
+}
+
+type MissingField struct {
+	Context ConversionContext
+	Path    Path
+	Value   any
+}
+
+type conversionErrors struct {
+	missingFields []missingFieldOnCopy
+}
+
+// VersionedObject wraps the standard GA, Alpha, Beta versions of a GCP
+// resource. By accessing the object using Access(), AccessAlpha().
+// AccessBeta(), VersionedObject will ensure that common fields between the
+// versions of the object are in sync.
+type VersionedObject[GA any, Alpha any, Beta any] struct {
+	ga    GA
+	alpha Alpha
+	beta  Beta
+
+	errors [conversionContextCount]conversionErrors
+}
+
+// cycleCheck that there are no cycles where a struct type appears 2+ times on
+// the same path. Our algorithms requires special handling for recursive
+// structures.
+func cycleCheck(p Path, t reflect.Type, seen []string) error {
+	switch t.Kind() {
+	case reflect.Slice:
+		return cycleCheck(p.Index(0), t.Elem(), seen)
+	case reflect.Pointer:
+		return cycleCheck(p.Pointer(), t.Elem(), seen)
+	case reflect.Map:
+		// Use "x" as the placeholder for the map key in the Path for debugging
+		// output purposes.
+		return cycleCheck(p.MapIndex("x"), t.Elem(), seen)
+	case reflect.Struct:
+		typeName := fmt.Sprintf("%s/%s", t.PkgPath(), t.Name())
+		for _, seenTypeName := range seen {
+			if typeName == seenTypeName {
+				return fmt.Errorf("recursive type found at %s: %s", p, typeName)
+			}
+		}
+		// Add this struct type to the list of types seen on this path.
+		seen = append(seen, fmt.Sprintf("%s/%s", t.PkgPath(), t.Name()))
```

Comment:
pop the stack after processing?

Reply:
Is this needed? seen gets thrown away when the function returns...